gk/fib.c

```diff
@@ -1795,37 +1798,56 @@ list_ipv4_fib_entries(lua_State *l, struct gk_lpm *ltbl)
 	ret = rte_lpm_iterator_state_init(ltbl->lpm, 0, 0, &state);
 	if (ret < 0) {
 		rte_spinlock_unlock_tm(&ltbl->lock);
-		luaL_error(l, "gk: failed to initialize the lpm rule iterator state at %s!",
+		luaL_error(l, "gk: failed to initialize the lpm rule iterator state at %s",
 			__func__);
 	}
 
 	index = rte_lpm_rule_iterate(&state, &re4);
 	while (index >= 0) {
+		unsigned int num_addrs;
+		size_t dentry_size;
+
 		fib = &ltbl->fib_tbl[re4->next_hop];
 		if (fib->action == GK_FWD_NEIGHBOR_FRONT_NET ||
 				fib->action == GK_FWD_NEIGHBOR_BACK_NET) {
 			index = rte_lpm_rule_iterate(&state, &re4);
 			continue;
 		}
 
-		memset(&dentry, 0, sizeof(dentry));
-		dentry.addr.proto = RTE_ETHER_TYPE_IPV4;
-		dentry.addr.ip.v4.s_addr = htonl(re4->ip);
-		dentry.prefix_len = state.depth;
-		fillup_gk_fib_dump_entry(&dentry, fib);
+		num_addrs = fib->action == GK_FWD_GRANTOR
+			? fib->u.grantor.set->num_entries
+			: 1;
+		dentry_size = sizeof(*dentry) +
+			num_addrs * sizeof(*dentry->addr_sets);
+
+		dentry = rte_malloc("fib4_dump", dentry_size, 0);
+		if (dentry == NULL) {
+			rte_spinlock_unlock_tm(&ltbl->lock);
+			luaL_error(l,
+				"gk: failed to allocate memory for the IPv4 FIB dump at %s",
+				__func__);
+		}
+		memset(dentry, 0, dentry_size);
```

Comment:
Replacing `rte_malloc()` with `rte_zmalloc()` allows us to drop the `memset()`.

This comment applies to the IPv6 case as well.

Notice that we don't need `rte_zmalloc_socket()` here because the memory is not used by the GK blocks. This observation is worth a comment to help readers.

Reply:
Use `unlikely()` on the expression `dentry == NULL`.